### PR TITLE
ch5 issue69

### LINKED ISSUE #69: How should we choose the appropriate data splitting strategy to mitigate bias and leakage and to improve fairness?
**Section Outline**
Please list the key points of content you intend to cover for this section

1. What are the various data splits strategies (Static, Cross-validation, Stratified, Temporal)
2. Potential scenarios of bias in data splits and suggestions to remedy
3. Potential scenarios of unfairness in data splits and suggestions to remedy
4. Potential scenarios of data leakage in data splits and suggestions to remedy
 
**Definition of Done**
- [ ] Completed `5-data-mgmt-exp-proc/<filename.md>`
- [ ] Verified against [self-review checklist](https://github.com/aisingapore/ai-practitioner-handbook/blob/main/REVIEWING.md)
- [ ] Added section to [table of contents](https://github.com/aisingapore/ai-practitioner-handbook/blob/main/_toc.yml)
- [ ] Page builds and displays correctly on local machine

### PR DESCRIPTION
**Pull request by chuawjk**
Opening a new PR for review.

### CODE REVIEW

**Comment by chuawjk** [comment]
This PR covers two sections:
- [What are the various data split strategies?](https://aisingapore.github.io/handbook-staging/book/5-data-mgmt-exp-proc/data-splits.html)
- [Potential scenarios of bias, unfairness, data leakage in data splits and suggestions to remedy](https://aisingapore.github.io/handbook-staging/book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.html)

**Comment by chuawjk** [comment]
I'd like to invite @kevinosl @najibninaba @ngkh2018 to review this contribution by @joshxinjie.

Thanks Xin Jie! FYI, I opened a new PR for this as the previous one got automatically closed due to the automerge. Pls bear with me as I continue to try to optimise the contribution workflow.

**Comment by chuawjk** [comment]
Updated links on the staging site
https://oss.gitlab.aisingapore.net/aisg-handbook/ai-practitioner-handbook/book/5-data-mgmt-exp-proc/data-splits.html
https://oss.gitlab.aisingapore.net/aisg-handbook/ai-practitioner-handbook/book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.html

File: book/5-data-mgmt-exp-proc/data-splits.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+### Test Set
+The test set is the second hold-out set that will not be used during training. Since the validation set is commonly used to influence the modelling decisions, our ML model pipeline is likely to be biassed in performing well on the validation set. Hence, we will need a completely unseen set of data to provide a final benchmark of our model’s performance on unseen data. It is important to not use this test set to influence your modelling decision. Otherwise you will get an inflated sense of how well the model will perform in real life. You should only evaluate your model on the test set during the final step of the pipeline. As such, it is common for the train set performance to be higher than the validation performance, which in turn, should be higher than the test performance.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits.md:16]: Suggest adding a separate tip at the end:
```Tip: After finding the best model from validation, it is also good practice to re-run this model (with the best hyperparameters) on the full dataset (train + validation + test) and save the resulting weights for further use.```

File: book/5-data-mgmt-exp-proc/data-splits.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+One commonly used rule of thumb will be to adopt a 70-20-10 split for your train, validation and test sets respectively. Ultimately, you should prioritise selecting a ratio catered to your needs.
+
+
+## 2. Cross Validation
+A single, static validation set could potentially present a biassed assessment of model performance. This is particularly the case with smaller datasets where favourable validation performance may arise by chance. Cross-validation is an alternative to having a static validation set. In cross-validation, you will conduct multiple rounds of model training, each time with a different section of your dataset serving as the validation set. This will minimise the variance associated with choosing any particular choice of validation set. However, cross-validation does come with the penalty of additional training time. So it is usually recommended when you are dealing with a smaller dataset.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits.md]: ```... with smaller datasets where there could be insufficient samples to form a reliable validation set. In such cases, favourable validation performance may arise by chance due to a higher proportion of 'easier' examples appearing in the validation set.```

```... penalty of additional training time. The difference could be significant especially when training large models.```

Edited as above.

Also suggest adding another tip at the end:
```Tip: In cases of extremely small datasets, you could consider _nested cross-validation_ to maximise the usage of the whole dataset for training, validation and testing.```

File: book/5-data-mgmt-exp-proc/data-splits.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+
+## 3. Stratified Split
+For datasets with unbalanced distribution of targets and/or features, you may want to consider stratified splitting. Stratified splitting aims to split your dataset, while maintaining similar proportions of any desired features/targets across your train, validation and test sets. 
+
+An example of a dataset with features that are skewed in their distribution will be a credit worthiness classification problem, where you may have fewer individuals in the dataset with ages of less than 20 years old.  If it is critical to model the behaviour of this age group, you can stratify the dataset based on age buckets.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits.md]: ```... with ages of less than 20 years old. A random split could result in insufficient numbers of them getting assigned to the validation and test sets.```

```... , you can stratify the dataset based on age buckets. Doing so would fix the proportion of each age group getting assigned to each set.```

Edited as above.

File: book/5-data-mgmt-exp-proc/data-splits.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+An example of a dataset with imbalanced targets will be a fraud detection dataset, where fraudulent examples are typically the minority. In this case, you will stratify the dataset based on the target.
+
+In general, the larger the dataset, the less likely features and targets will be unevenly distributed. You should still check your dataset for any uneven distribution in features and targets.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits.md]: ```... less likely features and targets will be unevenly distributed across the sets.```
Edited as above.

**Comment by kevinosl** [comment]
@joshxinjie I've submitted my review for data-splits. Thanks!
PS: the links to the references seem to be missing.

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Potential scenarios of bias, unfairness, data leakage in data splits and suggestions to remedy
+
+Contributor(s): Lee Xin Jie
+
+
+## Potential scenarios of bias and unfairness in data splits and suggestions to remedy
+
+Having biases in datasets may lead to unfairness in our models. We want to ensure that model predictions are fair for different groups of users and scenarios.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md:8]: Suggesting adding the text below, after the 1st para:

---
There are 2 sources of such bias and unfairness:
i) The full dataset is skewed to certain subgroups (eg. we are training a model to predict voting behaviour in Singapore but the dataset contains 95% Chinese).
ii) The splitting of the full dataset into train, validation and test sets introduces skewness (eg. the train set contains 30% aged above 60 while the validation set contains 15% aged above 60).

This article discusses point ii). 

_Point i) is related to ensuring a representative dataset is curated at the start of the project, and this is partly discussed in [INSERT LINK TO THE PAGE 'What are the key areas to look out for in the data when framing the AI project?']._

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+When dealing with datasets involving geographical and demographical slices, you may also want to consider stratified data split to ensure that each split contains a balanced set of examples across geographical and demographical attributes. This will help reduce any potential unfairness in model predictions between users from different geographical and demographical slices.
+
+
+## Potential scenarios of data leakage in data splits and suggestions to remedy
+
+Data leakage is the scenario where the model has access to information that will not be available during production. Often, if your model performs surprisingly well on the validation/test set (e.g. 100% test set performance), you should perform a check for data leakages. Data leakages should be avoided, as they allow the model to leverage on leaked information and achieve inflated performance on the validation/test set.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: ```... is the scenario where during training, the model has access to information that will not be available during production.```

```... to leverage on leaked information during training and achieve inflated ...```

Edited as above.

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+A common example of data leakage is temporal data leakage, which may occur when you are building a model to forecast future events based on past events. For example, you are building a model to predict the price of a stock in future, and you have past daily data of the stock’s prices. If you perform a random data split, it is likely that the train set may contain future stock prices, while the test set may contain past stock prices. This will allow the model to leverage on future stock prices it encounters during training to predict past prices on the test set, thereby inflating the model’s performance.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: Suggest bold font for **temporal data leakage**

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+A common example of data leakage is temporal data leakage, which may occur when you are building a model to forecast future events based on past events. For example, you are building a model to predict the price of a stock in future, and you have past daily data of the stock’s prices. If you perform a random data split, it is likely that the train set may contain future stock prices, while the test set may contain past stock prices. This will allow the model to leverage on future stock prices it encounters during training to predict past prices on the test set, thereby inflating the model’s performance. 
+
+In temporal problems involving high correlations between successive times, you should also aim to group successive days together in the same data split to minimise data leakage.
+
+Datasets with duplicated data points are also likely to cause data leakages. If a random data split is adopted, you may have the same data points in both the train and test set. This will result in inflated test performance.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: Suggest bold font for **duplicated data points**

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+Datasets with near duplicates can also subtly introduce data leakage. One example will be in computer vision, where it is possible to have multiple images that are near duplicates but not perfect duplicates. For example, if you are training a model to classify car models, you may have images of the exact same car taken 1 second apart. In this case the images are likely to have very similar characteristics. If you split the dataset randomly, you are likely to split these images into each of the train, validation and test set. Your model will then likely evaluate well on the images assigned to the test set, since it has already encountered similar images during training. In this scenario, you should consider a stratified splitting of your dataset, where images of the same car should be assigned to a single dataset split.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: Suggest bold font for **near duplicates**

```..., where it is possible to have multiple images that are near duplicates but not perfect duplicates. This can easily occur when the data comes from video streams made up of continuous sequences of images.```

```..., where images of the same car should be assigned to a single dataset split. You could use timestamps or image hashing to achieve this.```

Edited as above.

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+When engineering features, ensure that you do not inject data leakage during the process. If you are building a model to predict how many items a user will purchase at a store on a particular day, you will want to only engineer features that are available during inference in production. Examples of features leveraging information that won’t be available are the number of items sold in the store during the same day, or a moving average of the number of items purchased by the user up to and including the day itself, etc.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: Suggest bold font for **engineering features**

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+Having a repeatable data split algorithm is important to avoid introducing unforeseen data leakages during model retraining. Imagine the scenario where an unrepeatable random data split algorithm is adopted. When you have new datasets, you will likely generate new data splits and trigger model retraining. After retraining, you will often have to compare the performance of the old model vs the new model, to choose which model to deploy. An unrepeatable data split algorithm may mean that some of the old training data may fall into the new validation dataset. Hence, your old model may appear to perform better on the new validation dataset and you will deploy that model. However you may soon realise that the actual model performance in production is lower after deployment.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md]: Suggest bold font for **repeatable data split algorithm**

File: book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+
+## References
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/data-splits-bias-fairness-leakage.md:40]: The links to the references seem to be missing.

**Comment by kevinosl** [comment]
@joshxinjie I've submitted my review for data-splits-bias-fairness-leakage. Thanks!
